File: src/main/java/thedarkcolour/futuremc/container/ContainerLoom.java
package thedarkcolour.futuremc.container;

import net.minecraft.client.gui.inventory.GuiContainer;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.entity.player.EntityPlayerMP;
import net.minecraft.entity.player.InventoryPlayer;
import net.minecraft.inventory.Slot;
import net.minecraft.item.EnumDyeColor;
import net.minecraft.item.ItemBanner;
import net.minecraft.item.ItemStack;
import net.minecraft.nbt.NBTTagCompound;
import net.minecraft.nbt.NBTTagList;
import net.minecraft.tileentity.BannerPattern;
import net.minecraft.util.math.BlockPos;
import net.minecraft.world.World;
import net.minecraftforge.fml.relauncher.Side;
import net.minecraftforge.fml.relauncher.SideOnly;
import net.minecraftforge.items.ItemStackHandler;
import net.minecraftforge.items.SlotItemHandler;
import thedarkcolour.core.gui.Container;
import thedarkcolour.futuremc.client.gui.GuiLoom;
import thedarkcolour.futuremc.compat.oredict.OreDict;
import thedarkcolour.futuremc.init.Init;
import thedarkcolour.futuremc.item.ItemBannerPattern;

import javax.annotation.Nonnull;

public class ContainerLoom extends Container {
    private InventoryPlayer playerInv;
    private World world;
    private BlockPos pos;
    private int selectedIndex;
    private Runnable inventoryUpdateListener = () -> {};
    private ItemStackHandler handler = new ItemStackHandler(4) {
        @Override
        protected void onContentsChanged(int slot) {
            if (slot != 3) {
                handleCrafting();
            }
            inventoryUpdateListener.run();
        }
    };

    public ContainerLoom(InventoryPlayer playerInv, World worldIn, BlockPos pos) {
        this.playerInv = playerInv;
        this.world = worldIn;
        this.pos = pos;

        addOwnSlots();
        addPlayerSlots();
    }

    private void addOwnSlots() {
        addSlotToContainer(new SlotItemHandler(handler, 0, 13, 26) {
            @Override
            public boolean isItemValid(@Nonnull ItemStack stack) {
                return stack.getItem() instanceof ItemBanner;
            }
        });
        addSlotToContainer(new SlotItemHandler(handler, 1, 33, 26) {
            @Override
            public boolean isItemValid(@Nonnull ItemStack stack) {
                return OreDict.getOreName(stack).startsWith("dye");
            }
        });
        addSlotToContainer(new SlotItemHandler(handler, 2, 23, 45) {
            @Override
            public boolean isItemValid(@Nonnull ItemStack stack) {
                return stack.getItem() instanceof ItemBannerPattern;
            }
        });
        addSlotToContainer(new SlotItemHandler(handler, 3, 143, 57) {
            @Override
            public boolean isItemValid(@Nonnull ItemStack stack) {
                return false;
            }

            @Override
            public ItemStack onTake(EntityPlayer thePlayer, ItemStack stack) {
                getColor().shrink(1);
                getBanner().shrink(1);
                return stack;
            }
        });
    }

    private void addPlayerSlots() {
        for (int row = 0; row < 3; ++row) {
            for (int col = 0; col < 9; ++col) {
                int x = col * 18 + 8;
                int y = row * 18 + 84;
                this.addSlotToContainer(new Slot(this.playerInv, col + row * 9 + 9, x, y));
            }
        }

        for (int row = 0; row < 9; ++row) {
            int x = 9 + row * 18 - 1;
            int y = 58 + 70 + 14;
            this.addSlotToContainer(new Slot(this.playerInv, row, x, y));
        }
    }

    public void setColor(ItemStack stack) {
        handler.setStackInSlot(1, stack);
    }

    public void setPattern(ItemStack stack) {
        handler.setStackInSlot(2, stack);
    }

    public void setOutput(ItemStack stack) {
        handler.setStackInSlot(3, stack);
    }

    public ItemStack getBanner() {
        return handler.getStackInSlot(0);
    }

    public ItemStack getColor() {
        return handler.getStackInSlot(1);
    }

    public ItemStack getPattern() {
        return handler.getStackInSlot(2);
    }

    public ItemStack getOutput() {
        return handler.getStackInSlot(3);
    }

    public Slot getLoomSlot(int index) {
        return getSlot(index);
    }

    public int getSelectedIndex() {
        return selectedIndex;
    }

    @Override
    public boolean canInteractWith(EntityPlayer playerIn) {
        if (world.getBlockState(pos).getBlock() != Init.LOOM) {
            return false;
        } else {
            return playerIn.getDistanceSq((double)pos.getX() + 0.5D, (double)pos.getY() + 0.5D, (double)pos.getZ() + 0.5D) <= 64.0D;
        }
    }

    @Override
    public boolean enchantItem(EntityPlayer playerIn, int id) {
        if (id > 0 && id <= BannerPattern.values().length) {
            selectedIndex = id;
            updateRecipeResultSlot();
            return true;
        }

        return false;
    }

    private void handleCrafting() {
        ItemStack banner = getBanner();
        ItemStack pattern = getPattern();
        if (getOutput().isEmpty() || !banner.isEmpty() && !getColor().isEmpty() && selectedIndex > 0 && (selectedIndex < BannerPattern.values().length - 5 || !pattern.isEmpty())) {
            if (!pattern.isEmpty() && pattern.getItem() instanceof ItemBannerPattern) {
                NBTTagCompound tag = banner.getOrCreateSubCompound("BlockEntityTag");
                boolean flag = tag.hasKey("Patterns", 9) && !banner.isEmpty() && tag.getTagList("Patterns", 10).tagCount() >= 6;
                if (flag) {
                    selectedIndex = 0;
                } else {
                    selectedIndex = ItemBannerPattern.getBannerPattern(pattern).ordinal();
                }
            }
        } else {
            setOutput(ItemStack.EMPTY);
            selectedIndex = 0;
        }

        updateRecipeResultSlot();
        detectAndSendChanges();
    }

    private void updateRecipeResultSlot() {
        if (selectedIndex > 0) {
            ItemStack banner = getBanner();
            ItemStack color = getColor();
            ItemStack stack = ItemStack.EMPTY;
            if (!banner.isEmpty() && !color.isEmpty()) {
                stack = banner.copy();
                stack.setCount(1);
                NBTTagCompound nbt = stack.getOrCreateSubCompound("BlockEntityTag");
                NBTTagList list;
                if (nbt.hasKey("Patterns", 9)) {
                    list = nbt.getTagList("Pattern", 10);
                } else {
                    list = new NBTTagList();
                    nbt.setTag("Patterns", list);
                }
                NBTTagCompound tag = new NBTTagCompound();
                tag.setString("Pattern", BannerPattern.values()[selectedIndex].getHashname());
                tag.setInteger("Color", getColorForStack(color).getDyeDamage());
                list.appendTag(tag);
            }

            if (!ItemStack.areItemStacksEqual(stack, getOutput())) {
                setOutput(stack);
            }
        } else {
            setOutput(ItemStack.EMPTY);
        }
    }

    public void setInventoryUpdateListener(Runnable listenerIn) {
        inventoryUpdateListener = listenerIn;
    }

    @Override
    public ItemStack transferStackInSlot(EntityPlayer playerIn, int index) {
        ItemStack itemstack = ItemStack.EMPTY;
        Slot slot = inventorySlots.get(index);
        if (slot != null && slot.getHasStack()) {
            ItemStack itemstack1 = slot.getStack();
            itemstack = itemstack1.copy();
            if (index == 3) {
                if (!mergeItemStack(itemstack1, 4, 40, true)) {
                    return ItemStack.EMPTY;
                }

                slot.onSlotChange(itemstack1, itemstack);
            } else if (index != 1 && index != 0 && index != 2) {
                if (itemstack1.getItem() instanceof ItemBanner) {
                    if (!mergeItemStack(itemstack1, 0, 1, false)) {
                        return ItemStack.EMPTY;
                    }
                } else if (OreDict.getOreName(itemstack1).startsWith("dye")) {
                    if (!mergeItemStack(itemstack1, 1, 2, false)) {
                        return ItemStack.EMPTY;
                    }
                } else if (itemstack1.getItem() instanceof ItemBannerPattern) {
                    if (!mergeItemStack(itemstack1, 2, 3, false)) {
                        return ItemStack.EMPTY;
                    }
                } else if (index >= 4 && index < 31) {
                    if (!mergeItemStack(itemstack1, 31, 40, false)) {
                        return ItemStack.EMPTY;
                    }
                } else if (index >= 31 && index < 40 && !mergeItemStack(itemstack1, 4, 31, false)) {
                    return ItemStack.EMPTY;
                }
            } else if (!mergeItemStack(itemstack1, 4, 40, false)) {
                return ItemStack.EMPTY;
            }

            if (itemstack1.isEmpty()) {
                slot.putStack(ItemStack.EMPTY);
            } else {
                slot.onSlotChanged();
            }

            if (itemstack1.getCount() == itemstack.getCount()) {
                return ItemStack.EMPTY;
            }

            slot.onTake(playerIn, itemstack1);
        }

        return itemstack;
    }

    @Override
    public void onContainerClosed(EntityPlayer playerIn) {
        super.onContainerClosed(playerIn);
        if (!playerIn.isEntityAlive() || playerIn instanceof EntityPlayerMP && ((EntityPlayerMP) playerIn).hasDisconnected()) {
            for (int i = 0; i < 3; i++) {
                ItemStack stack = handler.getStackInSlot(i);
                if (!stack.isEmpty()) {
                    playerIn.entityDropItem(stack, 0.5F);
                }
            }
        } else {
            for (int i = 0; i < 3; i++) {
                if (!handler.getStackInSlot(i).isEmpty()) {
                    playerInv.placeItemBackInInventory(world, handler.getStackInSlot(i));
                }
            }
        }
    }

    public InventoryPlayer getPlayerInv() {
        return playerInv;
    }

    public static EnumDyeColor getColorForStack(ItemStack stack) {
        String s = OreDict.getOreName(stack);

        if(s.startsWith("dye")) {
            s = s.replaceFirst("dye", "");
            switch (s) {
                case "Black": return EnumDyeColor.BLACK;
                case "Red": return EnumDyeColor.RED;
                case "Green": return EnumDyeColor.GREEN;
                case "Purple": return EnumDyeColor.PURPLE;
                case "Cyan": return EnumDyeColor.CYAN;
                case "LightGray": return EnumDyeColor.SILVER;
                case "Gray": return EnumDyeColor.GRAY;
                case "Pink": return EnumDyeColor.PINK;
                case "Lime": return EnumDyeColor.LIME;
                case "Yellow": return EnumDyeColor.YELLOW;
                case "LightBlue": return EnumDyeColor.LIGHT_BLUE;
                case "Magenta": return EnumDyeColor.MAGENTA;
                case "Orange": return EnumDyeColor.ORANGE;
                default: return EnumDyeColor.WHITE;
            }
        } else {
            return EnumDyeColor.WHITE;
        }
    }

    @SideOnly(Side.CLIENT)
    public GuiContainer getGuiContainer() {
        return new GuiLoom(new ContainerLoom(playerInv, world, pos));
    }
}

File: src/main/java/thedarkcolour/futuremc/container/ContainerLoom.kt
package thedarkcolour.futuremc.container

import net.minecraft.entity.player.EntityPlayer
import net.minecraft.entity.player.EntityPlayerMP
import net.minecraft.entity.player.InventoryPlayer
import net.minecraft.inventory.Slot
import net.minecraft.item.EnumDyeColor
import net.minecraft.item.ItemBanner
import net.minecraft.item.ItemStack
import net.minecraft.nbt.NBTTagCompound
import net.minecraft.nbt.NBTTagList
import net.minecraft.tileentity.BannerPattern
import net.minecraft.util.math.BlockPos
import net.minecraft.world.World
import net.minecraftforge.fml.relauncher.Side
import net.minecraftforge.fml.relauncher.SideOnly
import thedarkcolour.core.gui.FContainer
import thedarkcolour.core.inventory.FInventory
import thedarkcolour.core.inventory.FInventorySlot
import thedarkcolour.core.util.anyMatch
import thedarkcolour.core.util.getOreNames
import thedarkcolour.core.util.janyMatch
import thedarkcolour.futuremc.client.gui.GuiLoom
import thedarkcolour.futuremc.item.BannerPatternItem
import thedarkcolour.futuremc.item.BannerPatternItem.Companion.getBannerPattern
import thedarkcolour.futuremc.registry.FBlocks.LOOM

class ContainerLoom(
    playerInv: InventoryPlayer,
    private val world: World,
    private val pos: BlockPos
) : FContainer(playerInv) {
    private val handler = object : FInventory(4) {
        override fun onContentsChanged(slot: Int) {
            if (slot != 3) {
                handleCrafting()
            }
            inventoryUpdateListener()
        }

        override fun isItemValid(slot: Int, stack: ItemStack): Boolean {
            return when (slot) {
                0 -> stack.item is ItemBanner
                1 -> isDye(stack)
                2 -> stack.item is BannerPatternItem
                else -> false
            }
        }

        override fun onTake(playerIn: EntityPlayer, stack: ItemStack, slot: Int): ItemStack {
            if (slot == 3) {
                color.shrink(1)
                banner.shrink(1)
            }
            return super.onTake(playerIn, stack, slot)
        }
    }
    private var selectedIndex = 0
    private var inventoryUpdateListener = {}
    private lateinit var bannerSlot: Slot
    private lateinit var dyeSlot: Slot
    private lateinit var patternSlot: Slot
    private lateinit var resultSlot: Slot

    init {
        addOwnSlots()
        addPlayerSlots(playerInv)
    }

    private fun addOwnSlots() {
        bannerSlot = addSlotToContainer(FInventorySlot(handler, 0, 13, 26))
        dyeSlot = addSlotToContainer(FInventorySlot(handler, 1, 33, 26))
        patternSlot = addSlotToContainer(FInventorySlot(handler, 2, 23, 45))
        resultSlot = addSlotToContainer(FInventorySlot(handler, 3, 143, 57))
    }

    // delegates
    val banner: ItemStack
        get() = handler[0]
    var color: ItemStack
        get() = handler[1]
        set(value) { handler[1] = value }
    var pattern: ItemStack
        get() = handler[2]
        set(value) { handler[2] = value }
    var output: ItemStack
        get() = handler[3]
        set(value) { handler[3] = value }

    fun getLoomSlot(index: Int): Slot {
        return when (index) {
            0 -> bannerSlot
            1 -> dyeSlot
            2 -> patternSlot
            3 -> resultSlot
            else -> throw IllegalArgumentException("Index $index must be in 0..3")
        }
    }

    override fun canInteractWith(playerIn: EntityPlayer): Boolean {
        return isBlockInRange(LOOM, world, pos, playerIn)
    }

    override fun enchantItem(playerIn: EntityPlayer, id: Int): Boolean {
        if (id in 0 .. BASIC_PATTERNS.size) {
            selectedIndex = id
            updateRecipeResultSlot()
            return true
        }
        return false
    }

    private fun handleCrafting() {
        val banner = banner
        val pattern = pattern
        if (output.isEmpty || !banner.isEmpty && !color.isEmpty && selectedIndex > 0 && (selectedIndex < BASIC_PATTERNS.size - 5 || !pattern.isEmpty)) {
            if (pattern.item is BannerPatternItem) {
                val tag = banner.getOrCreateSubCompound("BlockEntityTag")
                val flag = tag.hasKey("Patterns", 9) && !banner.isEmpty && tag.getTagList("Patterns", 10).tagCount() >= 6
                selectedIndex = if (flag) {
                    0
                } else {
                    getBannerPattern(pattern).ordinal
                }
            }
        } else {
            output = ItemStack.EMPTY
            selectedIndex = 0
        }

        if (pattern.isEmpty) {
            selectedIndex = 0
        }

        updateRecipeResultSlot()
        detectAndSendChanges()
    }

    private fun updateRecipeResultSlot() {
        if (selectedIndex > 0) {
            val banner = banner
            val color = color
            var stack = ItemStack.EMPTY
            if (!banner.isEmpty && !color.isEmpty) {
                stack = banner.copy()
                stack.count = 1
                val nbt = stack.getOrCreateSubCompound("BlockEntityTag")
                val list = if (nbt.hasKey("Patterns", 9)) {
                    nbt.getTagList("Patterns", 10)
                } else {
                    NBTTagList()
                }
                val tag = NBTTagCompound()
                var selectedIndex = selectedIndex
                val bannerPatterns = if (pattern.isEmpty) {
                    --selectedIndex
                    BASIC_PATTERNS
                } else {
                    BannerPattern.values()
                }
                tag.setString("Pattern", bannerPatterns[selectedIndex].hashname)
                tag.setInteger("Color", getColorForStack(color).dyeDamage)
                list.appendTag(tag)
                nbt.setTag("Patterns", list)
            }
            if (!ItemStack.areItemStacksEqual(stack, output)) {
                output = stack
            }
        } else {
            output = ItemStack.EMPTY
        }
    }

    fun setInventoryUpdateListener(listenerIn: () -> Unit) {
        inventoryUpdateListener = listenerIn
    }

    override fun transferStackInSlot(playerIn: EntityPlayer, index: Int): ItemStack {
        var itemstack = ItemStack.EMPTY
        val slot = inventorySlots[index]
        if (slot != null && slot.hasStack) {
            val itemstack1 = slot.stack
            itemstack = itemstack1.copy()
            if (index == 3) {
                if (!mergeItemStack(itemstack1, 4, 40, true)) {
                    return ItemStack.EMPTY
                }
                slot.onSlotChange(itemstack1, itemstack)
            } else if (index != 1 && index != 0 && index != 2) {
                if (itemstack1.item is ItemBanner) {
                    if (!mergeItemStack(itemstack1, 0, 1, false)) {
                        return ItemStack.EMPTY
                    }
                } else if (janyMatch(
                        getOreNames(itemstack1)
                    ) { oreName -> oreName.startsWith("dye") }
                ) {
                    if (!mergeItemStack(itemstack1, 1, 2, false)) {
                        return ItemStack.EMPTY
                    }
                } else if (itemstack1.item is BannerPatternItem) {
                    if (!mergeItemStack(itemstack1, 2, 3, false)) {
                        return ItemStack.EMPTY
                    }
                } else if (index in 4..30) {
                    if (!mergeItemStack(itemstack1, 31, 40, false)) {
                        return ItemStack.EMPTY
                    }
                } else if (index in 31..39 && !mergeItemStack(itemstack1, 4, 31, false)) {
                    return ItemStack.EMPTY
                }
            } else if (!mergeItemStack(itemstack1, 4, 40, false)) {
                return ItemStack.EMPTY
            }
            if (itemstack1.isEmpty) {
                slot.putStack(ItemStack.EMPTY)
            } else {
                slot.onSlotChanged()
            }
            if (itemstack1.count == itemstack.count) {
                return ItemStack.EMPTY
            }
            slot.onTake(playerIn, itemstack1)
        }
        return itemstack
    }

    override fun onContainerClosed(playerIn: EntityPlayer) {
        super.onContainerClosed(playerIn)
        if (!playerIn.isEntityAlive || playerIn is EntityPlayerMP && playerIn.hasDisconnected()) {
            for (i in 0..2) {
                val stack = handler.getStackInSlot(i)
                if (!stack.isEmpty) {
                    playerIn.entityDropItem(stack, 0.5f)
                }
            }
        } else {
            for (i in 0..2) {
                if (!handler.getStackInSlot(i).isEmpty) {
                    playerInv.placeItemBackInInventory(world, handler.getStackInSlot(i))
                }
            }
        }
    }

    @SideOnly(Side.CLIENT)
    override fun createGui(): Any {
        return GuiLoom(ContainerLoom(playerInv, world, pos))
    }

    fun getSelectedIndex(): Int {
        return selectedIndex
    }

    companion object {
        // Vanilla banner patterns that don't require a pattern to create
        val BASIC_PATTERNS = arrayOf(BannerPattern.SQUARE_BOTTOM_LEFT, BannerPattern.SQUARE_BOTTOM_RIGHT, BannerPattern.SQUARE_TOP_LEFT, BannerPattern.SQUARE_TOP_RIGHT, BannerPattern.STRIPE_BOTTOM, BannerPattern.STRIPE_TOP, BannerPattern.STRIPE_LEFT, BannerPattern.STRIPE_RIGHT, BannerPattern.STRIPE_CENTER, BannerPattern.STRIPE_MIDDLE, BannerPattern.STRIPE_DOWNRIGHT, BannerPattern.STRIPE_DOWNLEFT, BannerPattern.STRIPE_SMALL, BannerPattern.CROSS, BannerPattern.STRAIGHT_CROSS, BannerPattern.TRIANGLE_BOTTOM, BannerPattern.TRIANGLE_TOP, BannerPattern.TRIANGLES_BOTTOM, BannerPattern.TRIANGLES_TOP, BannerPattern.DIAGONAL_LEFT, BannerPattern.DIAGONAL_RIGHT, BannerPattern.DIAGONAL_LEFT_MIRROR, BannerPattern.DIAGONAL_RIGHT_MIRROR, BannerPattern.CIRCLE_MIDDLE, BannerPattern.RHOMBUS_MIDDLE, BannerPattern.HALF_VERTICAL, BannerPattern.HALF_HORIZONTAL, BannerPattern.HALF_VERTICAL_MIRROR, BannerPattern.HALF_HORIZONTAL_MIRROR, BannerPattern.BORDER, BannerPattern.CURLY_BORDER, BannerPattern.GRADIENT, BannerPattern.GRADIENT_UP, BannerPattern.BRICKS)

        fun isDye(stack: ItemStack): Boolean {
            return getOreNames(stack).anyMatch { it.startsWith("dye") }
        }

        fun getColorForStack(stack: ItemStack): EnumDyeColor {
            for (s in getOreNames(stack)) {
                if (s.startsWith("dye")) {
                    when (s) {
                        "dyeBlack" -> return EnumDyeColor.BLACK
                        "dyeBlue" -> return EnumDyeColor.BLUE
                        "dyeBrown" -> return EnumDyeColor.BROWN
                        "dyeRed" -> return EnumDyeColor.RED
                        "dyeGreen" -> return EnumDyeColor.GREEN
                        "dyePurple" -> return EnumDyeColor.PURPLE
                        "dyeCyan" -> return EnumDyeColor.CYAN
                        "dyeLightGray" -> return EnumDyeColor.SILVER
                        "dyeGray" -> return EnumDyeColor.GRAY
                        "dyePink" -> return EnumDyeColor.PINK
                        "dyeLime" -> return EnumDyeColor.LIME
                        "dyeYellow" -> return EnumDyeColor.YELLOW
                        "dyeLightBlue" -> return EnumDyeColor.LIGHT_BLUE
                        "dyeMagenta" -> return EnumDyeColor.MAGENTA
                        "dyeOrange" -> return EnumDyeColor.ORANGE
                    }
                }
            }
            return EnumDyeColor.WHITE
        }
    }
}